lib/WP_Auth0_Nonce_Handler.php

```diff
@@ -1,134 +1,180 @@
-    return setcookie( self::COOKIE_NAME, $this->_uniqid, time() + self::COOKIE_EXPIRES, '/' );
-  }
-
-  /**
-   * Reset the nonce cookie value
-   *
-   * @return bool
-   */
-  public function reset() {
-    return setcookie( self::COOKIE_NAME, '', 0 );
-  }
-
-  /**
-   * Generate a random ID
-   * If using on PHP 7, it will be cryptographically secure
-   *
-   * @see https://secure.php.net/manual/en/function.random-bytes.php
-   *
-   * @param int $bytes - number of bytes to generate
-   *
-   * @return string
-   */
-  public function generateNonce( $bytes = 32 ) {
-    $nonce_bytes = function_exists( 'random_bytes' ) ? random_bytes( $bytes ) : openssl_random_pseudo_bytes( $bytes );
-    return bin2hex( $nonce_bytes );
-  }
-}
\ No newline at end of file
+/**
+ * Contains WP_Auth0_Nonce_Handler.
+ *
+ * @package WP-Auth0
+ */
+
+/**
+ * Class WP_Auth0_Nonce_Handler for generating and storing nonce-type values.
+ */
+class WP_Auth0_Nonce_Handler {
+
+	/**
+	 * Cookie name used for storage and verification.
+	 *
+	 * @var string
+	 */
+	const UNIQID_COOKIE_NAME = 'auth0_nonce_uniqid';
+
+	/**
+	 * Time, in seconds, for the cookie to last.
+	 * Added to time() to determine expiration time.
+	 *
+	 * @var integer
+	 */
+	const COOKIE_EXPIRES = HOUR_IN_SECONDS;
+
+	/**
+	 * Singleton class instance.
+	 *
+	 * @var WP_Auth0_Nonce_Handler|null
+	 */
+	protected static $_instance = null;
+
+	/**
+	 * Unique ID used as a nonce or salting.
+	 *
+	 * @var string
+	 */
+	private $_uniqid;
+
+	/**
+	 * Private to prevent cloning.
+	 */
+	private function __clone() {}
+
+	/**
+	 * Private to prevent unserializing.
+	 */
+	private function __wakeup() {}
+
+	/**
+	 * WP_Auth0_Nonce_Handler constructor.
+	 * Private to prevent new instances of this class.
+	 */
+	private function __construct() {
+		$this->init();
+	}
+
+	/**
+	 * Start-up process to make sure we have something stored.
+	 */
+	private function init() {
+		if ( isset( $_COOKIE[ static::UNIQID_COOKIE_NAME ] ) ) {
+			// Have a cookie, don't want to generate a new one.
+			$this->_uniqid = $_COOKIE[ static::UNIQID_COOKIE_NAME ];
+		} else {
+			// No cookie, need to create one.
+			$this->_uniqid = $this->generate_nonce();
+		}
+	}
+
+	/**
+	 * Get the internal instance of the singleton.
+	 *
+	 * @return WP_Auth0_State_Handler|WP_Auth0_Nonce_Handler|WP_Auth0_Nonce_Handler
+	 */
+	final public static function get_instance() {
+		if ( is_null( static::$_instance ) ) {
+			static::$_instance = new static();
+		}
+		return static::$_instance;
+	}
+
+	/**
+	 * Return the unique ID used for validation.
+	 *
+	 * @return string
+	 */
+	public function get_uniqid() {
+		return $this->_uniqid;
+	}
+
+	/**
+	 * Return the cookie expiration time to set.
+	 *
+	 * @return integer
+	 */
+	public function get_cookie_exp() {
+		return time() + self::COOKIE_EXPIRES;
+	}
+
+	/**
+	 * Set the random storage cookie.
+	 *
+	 * @return bool
+	 */
+	public function set_cookie() {
```

Comment:
shouldn't this be called internally after generating a new nonce? (private). It doesn't receive any param so I don't see the point of it being public at all